src/node_report_module.cc

```diff
@@ -0,0 +1,315 @@
+void GetReport(const FunctionCallbackInfo<Value>& info) {
+  Environment* env = Environment::GetCurrent(info);
+  Isolate* isolate = env->isolate();
+  HandleScope scope(isolate);
+  std::ostringstream out;
+
+  GetNodeReport(isolate, env, "JavaScript API", __func__, info[0], out);
+
+  // Return value is the contents of a report as a string.
+  info.GetReturnValue().Set(String::NewFromUtf8(isolate, out.str().c_str()));
+}
+
+/*******************************************************************************
+ * Callbacks for triggering report on fatal error, uncaught exception and
+ * external signals
+ ******************************************************************************/
+static void OnFatalError(const char* location, const char* message) {
+  // Don't obtain the environment here, as we could be not in any context
+  // TODO(gireeshpunathil) when https://github.com/nodejs/node/pull/22819
+  // lands, revisit this - obtain Environment safely.
+  Isolate* isolate = Isolate::GetCurrent();
+  // Environment* env = Environment::GetCurrent(isolate);
+  // std::shared_ptr<PerIsolateOptions> options =
+  // env->isolate_data()->options();
+  if (location) {
+    fprintf(stderr, "FATAL ERROR: %s %s\n", location, message);
+  } else {
+    fprintf(stderr, "FATAL ERROR: %s\n", message);
+  }
+  // Trigger report if requested
+  // As current environment may not be usable or damaged, we
+  // cannot check whether this situation warrants a report generation
+  // based on any modified user options. However, if we reach this point
+  // means that user has requested for a report on fatalerrors in the past.
+  // so latch on that configuration, go ahead and produce one.
+  // TODO(gireeshpunathil) see if this behavior can be enhanced using
+  // storing the tunable in a safely accessible location.
+  // if (options->report_catch_fatalerror) {
+  TriggerNodeReport(
+      isolate, nullptr, message, __func__, nullptr, MaybeLocal<Value>());
+  fflush(stderr);
+  raise(SIGABRT);
+}
+
+void OnUncaughtException(const FunctionCallbackInfo<Value>& info) {
+  Environment* env = Environment::GetCurrent(info);
+  Isolate* isolate = env->isolate();
+  HandleScope scope(isolate);
+  std::string cmdline = node::per_process_opts->cmdline;
+  std::shared_ptr<PerIsolateOptions> options = env->isolate_data()->options();
+
+  // Trigger report if requested
+  if (options->report_catch_exception) {
+    TriggerNodeReport(isolate, env, "exception", __func__, nullptr, info[0]);
+  }
+}
+
+/*******************************************************************************
+ * Signal handler for report action, called from JS land (util.js)
+ ******************************************************************************/
+
+void OnUserSignal(const FunctionCallbackInfo<Value>& info) {
+  Environment* env = Environment::GetCurrent(info);
+  Isolate* isolate = env->isolate();
+  CHECK(info[0]->IsString());
```

Comment:
This can crash if a user performs e.g. `process.emit('SIGUSR2')` manually, so this should probably be accompanied by typechecks in JS

Reply:
if you remember our earlier conversation on this - control the signal handling from JS land so that every follows the libuv's native semantics (`uv_async_t`). So the signal is intercepted in JS callback first, and I am explicitly passing a string (signal number), and that is the only route to `OnUserSignal`. So I guess we are safe here?

Reply:
```js
$ ./node --report-catch-signal --experimental-report
> process.emit('SIGUSR2')
./node[18114]: ../src/node_report_module.cc:137:void nodereport::OnUserSignal(const v8::FunctionCallbackInfo<v8::Value>&): Assertion `info[0]->IsString()' failed.
 1: 0x56537a4976e8 node::Abort() [./node]
 2: 0x56537a4977ed  [./node]
 3: 0x56537a5d99a0 nodereport::OnUserSignal(v8::FunctionCallbackInfo<v8::Value> const&) [./node]
 4: 0x56537a7203b7  [./node]
 5: 0x56537a721761 v8::internal::Builtin_HandleApiCall(int, v8::internal::Object**, v8::internal::Isolate*) [./node]
 6: 0x20b57055005d 
Aborted (core dumped)
```

@gireeshpunathil You could call it unfortunate that “anybody” can emit events on an EventEmitter, not just the object itself, but generally crashing the process using public API is something we’d very much like to avoid.

Reply:
@addaleax - fixed in the JS land thus, PTAL

```patch
--- a/lib/util.js
+++ b/lib/util.js
@@ -555,6 +555,8 @@ if (experimentalReport) {
     }
 
     function handleSignal(signo) {
+      if (signo === undefined || signo == null)
+        signo = config.signal;
       nr.onUserSignal(signo);
     }
```

Reply:
I mean … I guess it’s getting into technicalities, but any `signo` value that isn’t a string would crash, right? `typeof signo !== 'string'` should be okay

Reply:
@addaleax - done.